codegen/smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGeneratorTest.kt

```diff
@@ -47,14 +47,19 @@ internal class SmokeTestOperationSerializer: HttpSerialize<SmokeTestRequest> {
         builder.method = HttpMethod.POST
 
         builder.url {
-            val pathSegments = listOf<String>(
-                "smoketest",
-                "$label1".encodeLabel(),
-                "foo",
-            )
-            path = pathSegments.joinToString(separator = "/", prefix = "/")
-            parameters {
-                if (input.query1 != null) append("Query1", input.query1)
+            path.encodedSegments {
+                add(PercentEncoding.Path.encode("smoketest"))
+                "$label1".split("/").mapTo(this) { PercentEncoding.SmithyLabel.encode(it) }
+                add(PercentEncoding.Path.encode("foo"))
+            }
+            parameters.encodedParameters {
+                val labels = mutableMultiMapOf<String, String>()
```

Comment:
**Question**: Why do we need an auxiliary data structure here?

Wonder if we can have an extension function or something that allows you to set the encoding to use.

```kotlin
        public fun encodedParameters(
            encoding: Encoding = PercentEncoding.Query,
            block: MutableMultiMap<String, String>.() -> Unit
        ) {
            asView(
                Encodable::encoded,
                encoding::encodableFromEncoded,
                Encodable::encoded,
                encoding::encodableFromEncoded,
            ).apply(block)
        }

```

```kotlin
parameters.encodedParameters(PercentEncoding.SmithyLabel) {
     if (input.query1 != null) add("Query1", input.query1)
} 
```